src/V3Width.cpp

```diff
@@ -2150,14 +2155,32 @@ class WidthVisitor final : public VNVisitor {
         // Make sure dtype is sized
         nodep->dtypep(iterateEditMoveDTypep(nodep, nodep->subDTypep()));
         UASSERT_OBJ(nodep->dtypep(), nodep, "No dtype determined for var");
-        if (const AstUnsizedArrayDType* const unsizedp
-            = VN_CAST(nodep->dtypeSkipRefp(), UnsizedArrayDType)) {
-            if (!(m_ftaskp && m_ftaskp->dpiImport())) {
-                UINFO(9, "Unsized becomes dynamic array " << nodep << endl);
-                AstDynArrayDType* const newp
-                    = new AstDynArrayDType{unsizedp->fileline(), unsizedp->subDTypep()};
-                nodep->dtypep(newp);
-                v3Global.rootp()->typeTablep()->addTypesp(newp);
+        if (m_ftaskp && m_ftaskp->dpiImport()) {
+            AstNodeDType *dtp = nodep->dtypep();
+            AstNodeDType *np = nullptr;
+            while (VN_IS(dtp->skipRefp(), DynArrayDType)
+                    || VN_IS(dtp->skipRefp(), UnpackArrayDType)) {
+                if (const AstDynArrayDType* const unsizedp
+                    = VN_CAST(dtp->skipRefp(), DynArrayDType)) {
+                    if (!np) {
+                        UINFO(9, "Dynamic becomes unsized array (var itself) " << nodep << endl);
+                    } else {
+                        UINFO(9, "Dynamic becomes unsized array (subDType) " << dtp << endl);
+                    }
+                    AstUnsizedArrayDType* const newp
+                        = new AstUnsizedArrayDType{unsizedp->fileline(), unsizedp->subDTypep()};
+                    newp->dtypep(newp);
+                    if (!np) { // for Var itself
+                        nodep->dtypep(newp);
+                    } else { // for subDType
+                        np->virtRefDTypep(newp);
+                    }
+                    v3Global.rootp()->typeTablep()->addTypesp(newp);
+                    np = newp;
+                } else {
+                    np = dtp->skipRefp();
+                }
+                dtp = np->virtRefDTypep();
```

Comment:
Removing transformation either requires new iteration in V3Task when checking for basic type or leads to chain of strange errors in strange places (if `basicp()` is provided for `DynArrayDType`).